napari/components/_layer_slicer.py

```diff
@@ -43,9 +46,68 @@ def __init__(self):
         """
         self.events = EmitterGroup(source=self, ready=Event)
         self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._force_sync = False
         self._layers_to_task: Dict[Tuple[Layer], Future] = {}
         self._lock_layers_to_task = RLock()
 
+    @contextmanager
+    def force_sync(self):
+        """Context manager to allow a temporary forced synchronous slice.
+        This method only holds the _force_sync variable as True while the
+        manager is open, then resets it back to False after the manager is
+        closed.
+        """
+        self._force_sync = True
+        yield None
+        self._force_sync = False
+
+    def await_slice(self, future: Future, timeout: float = 5) -> None:
+        """Wait for a single slicing task to complete
+
+        Attributes
+        ----------
+        future: concurrent.futures.Future
+            asynchronous slicing task
+        timeout: float
+            time in seconds to wait before raising TimeoutError, defaults to 5s
+
+        Raises
+        ------
+        TimeoutError: when the timeout limit has been exceeded and the task is
+            not yet complete
+        """
+        done, _ = wait([future], timeout=timeout)
+        if done:
+            return
+        else:
+            raise TimeoutError(
+                f'Slice task did not complete within timeout ({timeout}s).'
+            )
+
+    def wait_until_idle(self, timeout: Optional(float or None) = None) -> None:
+        """Wait for all slicing tasks to complete before returning.
+
+        Attributes
+        ----------
+        timeout: float or None
+            (Optional) time in seconds to wait before raising TimeoutError. If set as None,
+            there is no limit to the wait time. Defaults to None
+
+        Raises
+        ------
+        TimeoutError: when the timeout limit has been exceeded and the task is
+            not yet complete
+        """
+        futures = self._layers_to_task.values()
```

Comment:
I would have thought we'd need to acquire `_lock_layers_to_task` here, so that we block any new tasks from being added - that would be the expected behavior of this method for me anyway. Ah, but will that deadlock with `_on_slice_done`?

Reply:
Our assumption is that slice requests should only come from the main thread, so this should be fine because we also expect this to be called from the main thread, so new requests will not be added once we're in here.